api/interface.go

```diff
@@ -185,6 +185,11 @@ func NewLoginResultParams(result params.LoginResult) (*LoginResultParams, error)
 type LoginProvider interface {
 	// Login performs log in when connecting to the controller.
 	Login(ctx context.Context, caller base.APICaller) (*LoginResultParams, error)
+	// Token returns the string used for authentication. Some providers may only obtain
+	// a token after login has completed.
+	// This is normally used as part of basic authentication in scenarios where a client
+	// makes use of a StreamConnector like when fetching logs using `juju debug-log`.
+	Token() string
```

Comment:
this could be in a separate `BasicAuthTokenProvider` interface and only `LoginProviders` where this method makes sense would implement it and in `api/state.go`, `LoginToken` method we can check if a `LoginProvider` implements the `BasicAuthTokenProvider` interface

Reply:
also naming suggestion: s/Token/AuthToken/

Reply:
All login providers should have this method, it's needed for a client to use any of the streaming endpoints. The way I refactored the code, the connection state doesn't need to hold onto the password anymore, it simply uses the login provider's token.
`userPassLoginProvider` - Token returns password.
`sessionTokenLoginProvider` - Token returns the session token.
`clientCredentialsLoginProvider` - Token returns client secret.
`tryInOrderLoginProvider` - Token returns the underlying provider's token.
The name I'm happy to change.